hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystemV1.java

```diff
@@ -260,6 +260,44 @@ private void testListFilesRecursive() throws Exception {
             expectedFilesCount, actualCount);
   }
 
+
+  protected void testRenameDir() throws Exception {
+    final String root = "/root_dir";
+    final String dir = root + "/dir1";
+    final Path source = new Path(fs.getUri().toString() + dir);
+    final Path dest = new Path(source.toString() + ".renamed");
+    // Add a sub-dir to the directory to be moved.
+    final Path subdir = new Path(source, "sub_dir1");
+    fs.mkdirs(subdir);
+    LOG.info("Created dir {}", subdir);
+
+    // case-1) source is a sub-dir to destin
+    final Path sourceRoot = new Path(fs.getUri().toString() + root);
+    LOG.info("Rename op-> source:/root_dir to destin:/root_dir/dir1/sub_dir1");
+    try {
+      fs.rename(sourceRoot, subdir);
+      Assert.fail("Should throw exception : Cannot rename a directory to" +
+              " its own subdirectory");
+    } catch (OMException e) {
+      // expected
+    }
+
+    LOG.info("Will move {} to {}", source, dest);
+    fs.rename(source, dest);
+
+    assertTrue("Directory rename failed", fs.exists(dest));
+    // Verify that the subdir is also renamed i.e. keys corresponding to the
+    // sub-directories of the renamed directory have also been renamed.
+    assertTrue("Keys under the renamed directory not renamed",
+            fs.exists(new Path(dest, "sub_dir1")));
+
+    // Test if one path belongs to other FileSystem.
+    Path fakeDir = new Path(fs.getUri().toString() + "fake" + dir);
+    LambdaTestUtils.intercept(IllegalArgumentException.class, "Wrong FS",
+            () -> fs.rename(fakeDir, dest));
```

Comment:
Can we add additional test cases that can fully cover current rename logic:

* Rename to a existed file, failed.

* Rename to a existed folder:
Rename from /a, to /b, final key will be /b/a

 * Rename to a parent folder:
case1: parent folder does't exist, and then failed
case2: parent folder existed, then succeed

Reply:
@linyiqun  Thanks a lot for the review comments.

Sure, I will add tests. Can you please help me to understand about case1 with some example.
>case1: parent folder does't exist, and then failed
Rename from /a/b/c/d to /a. Are you talking about this? How would I simulate non-existent "/a", which is already a valid root of the source path, right?

>case2: parent folder existed, then succeed
I think you are saying that, Rename from "/a/b/c" to "/a". It will succeed and result will be "/a/c"

Reply:
>Rename from /a/b/c/d to /a. Are you talking about this? How would I simulate non-existent "/a", which is already a valid root of the source path, right?

Yes, you are right. Can we transform this to the case below:
dir: /a/b/c/d
file: /a/d, 
Rename from /a/b/c/d to /a, should be failed since /a/d is a existed file.

>I think you are saying that, Rename from "/a/b/c" to "/a". It will succeed and result will be "/a/c"

Yes, this is what I want to say.

Hope makes sense to you.

Reply:
Thanks for the clear explanation. Will prepare the changes!